drivers/vulkan/rendering_device_vulkan.cpp

```diff
@@ -6082,7 +6082,13 @@ Vector<uint8_t> RenderingDeviceVulkan::buffer_get_data(RID p_buffer, uint32_t p_
 	}
 
 	Buffer tmp_buffer;
+
+#ifdef __ANDROID__
+	_buffer_allocate(&tmp_buffer, p_size, VK_BUFFER_USAGE_TRANSFER_DST_BIT, VMA_MEMORY_USAGE_CPU_ONLY, VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT);
```

Comment:
Its nice to see this issue getting some attention! And great that you have identified a solution. However, I am a bit concerned about using the `VMA_MEMORY_USAGE_CPU_ONLY` bit as it is [deprecated](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#ggaa5846affa1e9da3800e3e78fae2305cca40bdf4cddeffeb12f43d45ca1286e0a5)

I have a feeling that ``VMA_MEMORY_USAGE_AUTO_PREFER_HOST`` is the correct option and we need to add another flag to the last parameter. The [docs](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/usage_patterns.html) seem to indicate ``VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT | VMA_ALLOCATION_CREATE_MAPPED_BIT``

Reply:
Okay, there could be something deeper going on here.

`VMA_MEMORY_USAGE_CPU_ONLY` guarantees that `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` is set. ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`` allows us to get away without flushing or invalidating mapped memory regions using [vkFlushMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkFlushMappedMemoryRanges.html) and [vkInvalidateMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkInvalidateMappedMemoryRanges.html) respectively.

After allocating the memory, we do a GPU flush, but we don't call either of those functions. I wonder if the core of the issue is that we aren't properly making the memory available for reading. `VMA_MEMORY_USAGE_CPU_ONLY` works around the issue by using `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`, but since its deprecated, we should probably implement an actual solution.

Reply:
https://github.com/godotengine/godot/issues/78715#issuecomment-1807297175


Another solution would be to assign `requiredFlags`. If it is to be parameterized, then I think significantly more adjustments are required.

https://github.com/godotengine/godot/blob/59457685c18e2d729eea50c751c11f049a7186f0/drivers/vulkan/rendering_device_vulkan.cpp#L1363

`allocInfo.requiredFlags = VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`

Or are there other solutions?

`VMA_MEMORY_USAGE_CPU_ONLY` 
> Obsolete, preserved for **backward compatibility.** Guarantees VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT and VK_MEMORY_PROPERTY_HOST_COHERENT_BIT.

`Vulkan API 1.2.231` on MacOS M1
`Vulkan API 1.1.128` on Android. Maybe it is generally necessary with `version 1.1`, does anyone have an old graphics card for example?

Reply:
Okay, another update that explains things a bit more. It seems that on Desktop devices HOST_VISIBLE is pretty much always paired with HOST_COHERENT (whether you request it or not). While on Android, you can have HOST_VISIBLE without HOST_COHERENT (so presumably, you need to ask for it)

https://vulkan.gpuinfo.org/listmemory.php?platform=android

So we have two possible alternative solutions:
1. Ask for ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`` on Android devices (by modifying ``_buffer_allocate()``
2. Actually use [vkFlushMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkFlushMappedMemoryRanges.html) and [vkInvalidateMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkInvalidateMappedMemoryRanges.html) instead of relying on ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT``

Reply:
> Or are there other solutions?

The other solution is to use flush/invalidate I guess. I'm not sure if we are guaranteed to have HOST_VISIBLE memory available that also has ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT``. If we are, then we can just make ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`` required whenever we read back from the GPU, if not then the solution will be more complex

Reply:
> Edit: Sorry, I wrote all three comments without refreshing my page,

I also missed some comments, Github often updates automatically, not today. In the next few days I'll try the first solution. 

Perhaps this function should also be adapted.
https://github.com/godotengine/godot/blob/59457685c18e2d729eea50c751c11f049a7186f0/drivers/vulkan/rendering_device_vulkan.cpp#L1399

```cpp
	allocInfo.flags = VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT;
	allocInfo.usage = VMA_MEMORY_USAGE_AUTO_PREFER_HOST;
```

Reply:
I'll try this first, the VMA description is easier for me to understand

--------------
[vmaInvalidateAllocation](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#gaaa8412919139ef413a4215ac6a290fae) -> Calls `vkInvalidateMappedMemoryRanges` 
_It needs to be called **before reading** from a mapped memory for memory types that are not HOST_COHERENT. Map operation doesn't do that automatically._

[vmaFlushAllocation](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#ga30c37c1eec6025f397be41644f48490f)  -> Calls `vkFlushMappedMemoryRanges`
_It needs to be called **after writing** to a mapped memory for memory types that are not HOST_COHERENT. Unmap operation doesn't do that automatically._